src/Spryker/Zed/SynchronizationBehavior/Persistence/Propel/Behavior/SynchronizationBehavior.php

```diff
@@ -270,40 +290,102 @@ protected function addGenerateKeyMethod()
             $keySuffix = sprintf('get%s()', $filter->filter($parameters['key_suffix_column']['value']));
         }
 
-        if (isset($parameters['store'])) {
-            $storeSetStatement = "\$syncTransferData->setStore(\$this->store);";
+        if ($keySuffix !== null) {
+            $referenceSetStatement = "\$syncTransferData->setReference(\$this->$keySuffix);";
         }
 
-        if (isset($parameters['locale'])) {
-            $localeSetStatement = "\$syncTransferData->setLocale(\$this->locale);";
+        return "
+/**
+ * @return void
+ */
+protected function setGeneratedKey()
+{
+    \$syncTransferData = new \\Generated\\Shared\\Transfer\\SynchronizationDataTransfer();
+    $referenceSetStatement
+    $storeSetStatement
+    $localeSetStatement    
+    \$keyBuilder = \$this->getStorageKeyBuilder('$resource');
+
+    \$key = \$keyBuilder->generateKey(\$syncTransferData);
+    \$this->setKey(\$key);
+}        
+        ";
+    }
+
+    /**
+     * @throws \Spryker\Zed\SynchronizationBehavior\Persistence\Propel\Behavior\Exception\MissingAttributeException
+     *
+     * @return string
+     */
+    protected function addGenerateMappingResourceKeyMethod()
+    {
+        $parameters = $this->getParameters();
+        if (!isset($parameters['mapping_resource'])) {
+            return '/**
+ * @return void
+ */
+protected function setGeneratedKeyForMappingResource()
+{
+}';
+        }
+        $keySuffix = null;
+        $storeSetStatement = $this->getStoreStatement($parameters);
+        $localeSetStatement = $this->getLocaleStatement($parameters);
+        $referenceSetStatement = '';
+        $checkSuffixStatement = '';
+
+        if (!isset($parameters['resource']['value'])) {
+            throw new MissingAttributeException(sprintf(static::ERROR_MISSING_RESOURCE_PARAMETER, $this->getTable()->getPhpName()));
+        }
+
+        if (!isset($parameters['mapping_resource']['value'])) {
+            throw new MissingAttributeException(sprintf(static::ERROR_MISSING_MAPPING_RESOURCE_PARAMETER, $this->getTable()->getPhpName()));
+        }
+
+        $mappingResourceSuffix = "'{$parameters['mapping_resource']['value']}'";
+
+        $resource = $parameters['resource']['value'];
+
+        if (isset($parameters['mapping_resource_key_suffix_column'])) {
+            $filter = new UnderscoreToCamelCase();
+            $keySuffix = sprintf('get%s()', $filter->filter($parameters['mapping_resource_key_suffix_column']['value']));
+            $checkSuffixStatement = "if (empty(\$this->$keySuffix)) {
+         return;       
+    }
+            ";
         }
 
         if ($keySuffix !== null) {
-            $referenceSetStatement = "\$syncTransferData->setReference(\$this->$keySuffix);";
+            $referenceSetStatement = "\$syncTransferData->setReference($mappingResourceSuffix);";
```

Comment:
I think this is not necessary anymore, the next if will do the same assignment or if conditions is wrong.
Should be `$keySuffix === null`?